File: 733-flood-fill/733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        seen = set()
        q = deque([(sr, sc)])
        orig_color = image[sr][sc]
        
        while q:
            cur = q.popleft()
            x, y = cur
            seen.add((x, y))
            image[x][y] = color
            for dx, dy in [(1,0), (0,1), (-1,0), (0,-1)]:
                if (x+dx, y+dy) not in seen \
                and 0 <= x+dx < len(image) \
                and 0 <= y+dy < len(image[0]) \
                and image[x+dx][y+dy] == orig_color:
                    q.append((x+dx, y+dy))
        return image

File: turing_planet/bfs/733.flood-fill.cpp
/*
 * @lc app=leetcode id=733 lang=cpp
 *
 * [733] Flood Fill
 */

// @lc code=start
class Solution
{
  public:
#define VALID(i, j, maxi, maxj) ((i >= 0 && j >= 0 && i < maxi && j < maxj))

    vector<vector<int>> floodFill(vector<vector<int>> &image, int sr, int sc, int color)
    {
        int                  orig = image[sr][sc];
        int                  maxi = image.size(), maxj = image[0].size();
        vector<vector<bool>> visited(maxi, vector<bool>(maxj, false));

        queue<pair<int, int>> q;
        q.push({sr, sc});
        visited[sr][sc] = true;

        // mark start position as new color
        image[sr][sc] = color;
        while (!q.empty())
        {
            int qsize = q.size();
            while (qsize--)
            {
                auto [i, j] = q.front();
                q.pop();
                // iterate neighbors
                for (auto [newi, newj] :
                     vector<pair<int, int>>{{i + 1, j}, {i - 1, j}, {i, j + 1}, {i, j - 1}})
                {
                    if (VALID(newi, newj, maxi, maxj) && image[newi][newj] == orig &&
                        !visited[newi][newj])
                    {
                        visited[newi][newj] = true;
                        image[newi][newj]   = color;  // marked as used
                        q.push({newi, newj});
                    }
                }
            }
        }

        return image;
    }
};
// @lc code=end
